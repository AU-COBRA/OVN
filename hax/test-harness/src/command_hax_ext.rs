use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandHaxExt {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

/// Computes a list of arguments that setup the number of parallel
/// jobs for dune accordingly to environment variable `DUNEJOBS`.
fn dune_jobs_args() -> Vec<String> {
    if let Ok(jobs) = std::env::var("DUNEJOBS") {
        vec!["-j".into(), jobs]
    } else {
        vec![]
    }
}

impl CommandHaxExt for Command {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_hax: PathBuf,
        }
        const CARGO_HAX: &str = "cargo-hax";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built. Note this doesn't
                // include `hax-engine-names-extract`: its build
                // script requires the driver and CLI of `hax` to be
                // available.
                assert!(Command::new("cargo")
                            .args(&["build", "--bins"])
                            .current_dir(&root)
                            .status()
                            .unwrap()
                            .success());
                let cargo_hax = cargo_bin(CARGO_HAX);
                let driver = cargo_bin("driver-hax-frontend-exporter");
                // Now the driver & CLI are installed, call `cargo
                // build` injecting their paths
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bin", "hax-engine-names-extract"])
                        .env("HAX_CARGO_COMMAND_PATH", &cargo_hax)
                        .env("HAX_RUSTC_DRIVER_BINARY", &driver)
                        .current_dir(&root)
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .args(dune_jobs_args())
                        .env("HAX_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("hax-export-json-schemas"))
                        .env("HAX_ENGINE_NAMES_EXTRACT_BINARY", cargo_bin("hax-engine-names-extract"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver,
                    cargo_hax,
                    engine: engine_dir.join("_build/install/default/bin/hax-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_hax.clone());
                cmd.env("HAX_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("HAX_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_HAX),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
