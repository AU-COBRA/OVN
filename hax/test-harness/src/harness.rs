mod command_hax_ext;
use command_hax_ext::*;
use serde_json::{Map, Value};
use std::process::{Command, Stdio};

#[derive(Clone, Debug, serde::Serialize)]
pub enum TestKind {
    Translate { backend: String },
}

impl TestKind {
    fn as_name(&self) -> String {
        (match self {
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        })
        .join("-")
    }
}

#[allow(dead_code)]
fn bool_true() -> bool {
    true
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSnapshot {
    #[serde(default = "bool_true")]
    pub stderr: bool,
    #[serde(default = "bool_true")]
    pub stdout: bool,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// is the test optional? (useful for slow tests for instance)
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
    pub snapshot: TestSnapshot,
    pub include_flag: Option<String>,
    pub backend_options: Option<Vec<String>>,
}

impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        fn as_opt_bool(v: &Value, def: bool) -> Option<bool> {
            if v.is_null() {
                return Some(def);
            }
            v.as_bool()
        }
        fn as_bool(o: &Value, k: &str, def: bool) -> bool {
            let v = &o[k];
            as_opt_bool(v, def)
                .expect(format!("[{}] was expected to be a boolean, got {}", k, v).as_str())
        }
        let snapshot = &o["snapshot"];
        TestSpec {
            optional: as_bool(&o, "optional", false),
            broken: as_bool(&o, "broken", false),
            positive: as_bool(&o, "positive", true),
            issue_id: o["positive"].as_u64(),
            include_flag: o["include-flag"].as_str().map(|s| s.into()),
            backend_options: serde_json::from_value(o["backend-options"].clone()).unwrap(),
            snapshot: as_opt_bool(snapshot, true)
                .map(|b| TestSnapshot {
                    stderr: b,
                    stdout: b,
                })
                .or_else(|| match snapshot.as_str() {
                    Some(v @ ("stdout" | "stderr" | "both" | "none")) => Some(TestSnapshot {
                        stdout: matches!(v, "stdout" | "both"),
                        stderr: matches!(v, "stderr" | "both"),
                    }),
                    Some(v) => panic!(
                        "[snapshot] is \"{}\" but was expected to be \"stderr\", \"stdout\" or \"both\"", v
                    ),
                    None => None,
                })
                .unwrap_or_else(|| TestSnapshot {
                    stderr: as_bool(&snapshot, "stderr", true),
                    stdout: as_bool(&snapshot, "stdout", true),
                }),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug, serde::Serialize)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl Test {
    fn as_args(&self) -> Vec<String> {
        match &self.kind {
            TestKind::Translate { backend } => {
                let mut args = vec![];
                args.push("into".to_string());
                if let Some(i) = self.spec.include_flag.as_ref() {
                    args.push("-i".to_string());
                    args.push(i.to_string());
                }
                args.push("--dry-run".to_string());
                args.push(backend.clone());
                if let Some(backend_options) = &self.spec.backend_options {
                    args.extend_from_slice(backend_options.clone().as_slice());
                }
                args
            }
        }
    }
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::hax(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.as_args());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            use lazy_static::lazy_static;
            use regex::Regex;
            lazy_static! {
                // Regex [TIME] matches compile times
                static ref TIME: Regex = Regex::new(r"\bin \d+(\.\d+)?s\b").unwrap();
                static ref LOCK: Regex = Regex::new(r"Blocking waiting for \w+ lock on (the registry index|build directory|package cache)").unwrap();
            }
            TIME.replace_all(
                LOCK.replace_all(
                    &s.replace(r"\", "/").replace(&workspace, "WORKSPACE_ROOT"),
                    "",
                )
                .as_ref(),
                "in XXs",
            )
            .trim()
            .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = String::from_utf8_lossy(&out.stdout).to_string();

        // 3. make sure the test is successful
        let mut snapshot: Map<String, Value> = Map::new();
        if self.spec.snapshot.stderr {
            snapshot.insert("stderr".to_string(), Value::String(serr.clone()));
        }
        if self.spec.snapshot.stdout {
            snapshot.insert(
                "stdout".to_string(),
                serde_json::from_str(&sout)
                    .map(|out: hax_types::engine_api::Output| {
                        use serde_json::json;
                        json!({
                            "diagnostics": Value::Array(out.diagnostics.into_iter().map(|diag| json!({
                                "spans": Value::Array(diag.span.clone().into_iter().map(|span| Value::String(format!("{:?}", span))).collect()),
                                "message": Value::String(format!("{}", diag)),
                            })).collect()),
                            "files": Value::Object(out.files.into_iter().map(|file| (file.path, Value::String(file.contents))).collect())
                        })
                    })
                    .unwrap_or_else(|_| Value::String(cleanup(sout.clone()))),
            );
        }

        if !snapshot.is_empty() {
            let exit = out.status.code().unwrap_or(std::i32::MAX);
            snapshot.insert("exit".to_string(), exit.into());
            let snapshot = Value::Object(snapshot);
            let name = format!("{} {}", self.info.name, self.kind.as_name());

            let mut info = self.clone();
            info.info.manifest = info.info.manifest.strip_prefix(workspace).unwrap().into();
            insta::with_settings!({
                info => &info,
            }, { insta::assert_toml_snapshot!(name, snapshot) })
        }

        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (true, (true, false) | (false, true)) => Ok(()),
            (false, (false, false) | (true, true)) => Ok(()),
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.hax-tests], this function returns a list of tests
fn parse_hax_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    if metadata.is_null() {
        return vec![];
    }

    metadata
        .as_object()
        .expect(
            format!(
                "Expected value at key [hax-tests] to be a dictionary for package {:#?}",
                info
            )
            .as_str(),
        )
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(
                    format!(
                        "Expected value at key [{}] be a dictionary for package {:#?}",
                        a, info
                    )
                    .as_str(),
                )
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                _ => panic!(
                    "unexpected metadata [hax-tests.{}.{}] for package {:#?}",
                    a, b, info
                ),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    let mut args = libtest_mimic::Arguments::from_args();
    args.test_threads = Some(1);

    libtest_mimic::run(
        &args,
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_hax_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["hax-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
