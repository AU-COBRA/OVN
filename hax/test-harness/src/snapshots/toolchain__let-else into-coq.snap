---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: coq
  info:
    name: let-else
    manifest: let-else/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: false
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0

[stdout]
diagnostics = []

[stdout.files]
"Let_else.v" = '''
(* File automatically generated by Hacspec *)
From Hacspec Require Import Hacspec_Lib MachineIntegers.
From Coq Require Import ZArith.
Import List.ListNotations.
Open Scope Z_scope.
Open Scope bool_scope.

(*Not implemented yet? todo(item)*)

Definition let_else (opt : t_Option_t int32) : bool :=
  run match opt with
  | Option_Some x =>
    ControlFlow_Continue true
  | _ =>
    ControlFlow_Break false
  end.

Definition let_else_different_type (opt : t_Option_t int32) : bool :=
  run (let hoist1 := match opt with
    | Option_Some x =>
      ControlFlow_Continue (Option_Some (x.+(@repr WORDSIZE32 1)))
    | _ =>
      ControlFlow_Break false
    end : t_Option_t int32 in
  ControlFlow_Continue (let_else hoist1)).
'''
